pkg/web/api.go

```diff
@@ -79,19 +80,115 @@ func (api *API) RegisterControllers(controllers ...Controller) {
+// configured to match route.
+func (api *API) RegisterFilterBefore(beforeFilterName string, filter Filter) {
+	logrus.Debugf("Registering filter %s before %s", filter.Name(), beforeFilterName)
+	api.validateFilters(filter)
+	api.registerFilterRelatively(beforeFilterName, filter, func(beforeFilterPosition int) int {
+		return beforeFilterPosition
+	})
+}
+
+// RegisterFilterAfter registers the specified filter after the one with the given name.
+// If for some routes, the filter with the given name does not match the routes of the provided filter, then
+// the filter will be registered after the place at which the filter with this name would have been, had it been
+// configured to match route.
+func (api *API) RegisterFilterAfter(afterFilterName string, filter Filter) {
+	logrus.Debugf("Registering filter %s after %s", filter.Name(), afterFilterName)
+	api.validateFilters(filter)
+	api.registerFilterRelatively(afterFilterName, filter, func(filterPosition int) int {
+		return filterPosition + 1
+	})
+}
+
+// ReplaceFilter registers the given filter in the place of the filter with the given name.
+func (api *API) ReplaceFilter(replacedFilterName string, filter Filter) {
+	logrus.Debugf("Replacing filter %s with %s", replacedFilterName, filter.Name())
+	api.validateFilters(filter)
+	registeredFilterPosition := api.findFilterPosition(replacedFilterName)
+	api.Filters[registeredFilterPosition] = filter
+}
+
+// RemoveFilter removes the filter with the given name
+func (api *API) RemoveFilter(name string) {
+	position := api.findFilterPosition(name)
+	copy(api.Filters[position:], api.Filters[position+1:])
+	api.Filters[len(api.Filters)-1] = nil
+	api.Filters = api.Filters[:len(api.Filters)-1]
+}
+
 // RegisterPlugins registers a set of plugins
 func (api *API) RegisterPlugins(plugins ...Plugin) {
 	for _, plugin := range plugins {
-		pluginSegments := api.decomposePlugin(plugin)
-		if len(pluginSegments) == 0 {
-			logrus.Panicf("%T does not implement any plugin operation", plugin)
+		registeredFilterNames := api.filterNames(api.Filters)
+		if slice.StringsAnyPrefix(registeredFilterNames, plugin.Name()+":") {
+			logrus.Panicf("Plugin %s is already registered", plugin.Name())
 		}
+		pluginSegments := api.decomposePluginOrDie(plugin)
+		api.Filters = append(api.Filters, pluginSegments...)
+	}
+}
+
+func (api *API) validateFilters(filters ...Filter) {
+	newFilterNames := api.filterNames(filters)
+	if slice.StringsAnyEquals(newFilterNames, "") {
+		logrus.Panicf("Filters cannot have empty names")
+	}
+	registeredFilterNames := api.filterNames(api.Filters)
+	commonFilterNames := slice.StringsIntersection(registeredFilterNames, newFilterNames)
+	if len(commonFilterNames) > 0 {
+		logrus.Panicf("Filters %s are already registered", commonFilterNames)
+	}
+	if slice.StringsAnySubstring(newFilterNames, ":") {
+		logrus.Panic("Cannot register filters with : in their names")
```

Comment:
would it make sense to do the validation in way so that we may include the invalid name in the panic?